### ISSUE TITLE
onKeyboardItemSelected is called multiple times

### ISSUE THREAD

**Issue by Thorbenandresen** [wix/react-native-keyboard-input#6]
demoApp

**Repro**
1. add a console.log() to `onKeyboardItemSelected()`
1. Show keyboard by pressing "show2"
1. click in the text input
1. repeat the last two steps x times
1. Click "Click Me too!"

**Expected**
`onKeyboardItemSelected()` is only called once

**Actual**
`onKeyboardItemSelected()` is called x times

**Note:** 
Kinda hacky, but I currently work around it by debouncing  my method calls like this: 

```js
   this.latestStickerMessageTimeStamp = Date.now()
   [...]

  onKeyboardItemSelected(keyboardId, params) {

    if (keyboardId === 'KeyboardCustomStickers'){

      if (Date.now() - this.latestStickerMessageTimeStamp > 100) {
        params.stickerObj && this.onSubmitCustomKeyboardImage(params.stickerObj)
      }
      this.latestStickerMessageTimeStamp = Date.now()
}
```

**Comment by artald** [comment]
@Thorbenandresen interesting, I wasn't able to reproduce it.
Went back and forth between the text input and "show 2" keyboard, then pressed the "Click Me too!" button - got one print in my console. Tried to start with opening the "show 2" keyboard first, then tried a fresh start with the input first - still single console print.

Am I missing something?
Are you using the latest version of the package (3.0.3)? any chance you're using a different RN version or just what comes with the demo project (0.38)?

**Comment by jemise111** [comment]
Hey @artald this is actually happening to me as well. See the gif at the bottom for a repro.

I spend a few minutes digging around and couldn't find the cause but I noticed two things I just thought I'd throw out..

1) https://github.com/wix/react-native-keyboard-input/blob/master/src/CustomKeyboardView.js#L70 sometimes when this function is called `this.props.component` is `undefined` and `nextProps.component` is `undefined
2) https://github.com/wix/react-native-keyboard-input/blob/master/src/CustomKeyboardView.js implements `componentWillUnmount` twice. So I'm not sure if the first one is ever called

Hope that helps!

Repro:

![](http://g.recordit.co/iUHmVOYLFm.gif)

**Comment by artald** [comment]
Hey @jemise111 thanks for taking the time to look into it, and thanks for the repro. gif!

Regarding the points you've raised:
1. As far as I can tell, everything should still work fine if they are both `undefined`, but I'll need to take a closer look.
2. You are totally right, it was a mistake that I already fixed but didn't publish yet. Tough I doubt it if that's the reason for the multiple callbacks (the component is not unmounted in this scenario).

From the looks of it, it seems as if the component somehow registers multiple times for the `onItemSelected` event in the `KeyboardRegistry`, or fails to remove previous registers. I'll try to check it out when I can and see if I can find the cause. In the meantime, if you find any other clue, or if you're able to reproduce it on a different project (that is not the demo app) - please let me know. thanks!

**Comment by Thorbenandresen** [comment]
@artald thank you - yes, I can repro this in our chat app. I am using the custom keyboard as a sticker selector and the multiple callbacks result in multiple stickers being posted to the chat -
- unless I debounce the calls :)

**Comment by artald** [comment]
@Thorbenandresen cool, thanks for confirming.
Is there a specific scenario? Or just by switching to the stickers keyboard and selecting a sticker you get multiple callbacks?

**Comment by Thorbenandresen** [comment]
@artald the repro is what I described in my initial comment and/or what @jemise111 is describing. If I open the keyboard for the first time, it will only post one sticker.

**Comment by artald** [comment]
Thanks for the help guys.
A fix is available in `v3.0.5`. If you notice anything weird or find another scenario - please re-open.

**Comment by Thorbenandresen** [comment]
@artald that seems to have fixed it! Thanks for all the great work!
